### ISSUE TITLE
[Bug] - MSAL fails on deployed Xamarin Forms Android application with exception message

### ISSUE THREAD

**Issue by ryanwilson-gpdgroup** [AzureAD/microsoft-authentication-library-for-dotnet#2935]
I don't currently have any logs, it works fine in the android emulator, but when released to an android device via Intune Company Portal, it fails with the following: "**[Android Broker] The broker redirect uri is incorrect, it should be....**"

**Which version of MSAL.NET are you using?** **MSAL Version 4.35.0**

**Platform**
Xamarin Forms Android built with C# on .Net 5

**What authentication flow has the issue?**
* **Desktop / Mobile**
    **[X] Interactive**
    

Other?
I have provided a detailed explanation and screenshots of configurations on Stack Overflow at: https://stackoverflow.com/questions/69351967/xamarin-forms-android-authentication-failure-on-deployment-using-msal-library

**Is this a new or existing app?** This is a new app or experiment.
-->

**Repro**

```csharp
 //OAuthSettings is a class containing my values to pass to the methods of the 
            //PublicClientApplicationBuilder
            var builder = PublicClientApplicationBuilder
                                .Create(OAuthSettings.ApplicationId)
                                .WithTenantId(OAuthSettings.TenantId)
                                .WithBroker()
                                .WithRedirectUri(OAuthSettings.RedirectUri);

            PCA = builder.Build();
   
            try
            {
                var accounts = await PCA.GetAccountsAsync();

                var silentAuthResult = await PCA
                    .AcquireTokenSilent(new string[] { "api://xxxxxxxxxxxxxx/.default" }, accounts.FirstOrDefault())
                    .ExecuteAsync();

                AccessToken = new JwtSecurityToken(silentAuthResult.AccessToken);

                //more code removed for brevity
            }
            catch (MsalUiRequiredException msalEx)
            {
                
                var windowLocatorService = DependencyService.Get<IParentWindowLocatorService>();

                // Prompt the user to sign-in
                var interactiveRequest = PCA.AcquireTokenInteractive(new string[] { "api://xxxxxxxxxxxxxxxxxxx/.default" });

                //Used for Android and iOS
                AuthUIParent = windowLocatorService?.GetCurrentParentWindow();

                if (AuthUIParent != null)
                {
                    interactiveRequest = interactiveRequest
                        .WithParentActivityOrWindow(AuthUIParent);
                }
                //

                var interactiveAuthResult = await interactiveRequest.ExecuteAsync();

                AccessToken = new JwtSecurityToken(interactiveAuthResult.AccessToken);
           }
```

**Expected behavior**
I expect it to present the user with the login page to enter their email address and password, then to be prompted to authenticate using the Microsoft Authenticator as brokered authentication flow.

**Actual behavior**
As soon as the Login button is clicked to start the authentication process an error is presented with the following:
**[Android Broker] The broker redirect uri is incorrect, it should be....**

**Possible solution**
n/a

**Additional context / logs / screenshots**
Please see provided Stack Overflow link above for screenshots and more information

**Comment by bgavrilMS** [comment]
Feel free to open an issue on the Azure Portal, as I agree with you that they provide an experience that is not intuitive for redirect URIs.

**Comment by trwalke** [comment]
There are 2 redirect uris used for android @ryanwilson-gpdgroup. One for broker and one when the broker is not used. If you are using the system browser/embedded webview and are not using the broker you can use the following format:

`<data android:scheme="auth" android:host="msal{clientid}" />`

However, when you are using broker (Microsoft Authenticator/Intune Company Portal), you can use this format:

`<data android:scheme="msauth" android:host="{package name}" android:path="/{signature hash}" />`

These are critical for the Android OS to return to the MSAL client app after authentication with the system browser and broker.

We recommend developers add a manifest entry for both in order to ensure MSAL is always returned to.

See https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-xamarin-android-considerations#update-the-android-manifest-for-system-webview-support

and https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-xamarin-android-considerations#use-system-webview-in-brokered-authentication

Manifest example here:

```JSON

	<application android:theme="@android:style/Theme.NoTitleBar">
		<activity android:name="microsoft.identity.client.BrowserTabActivity" android:configChanges="orientation|screenSize">
			<intent-filter>
				<action android:name="android.intent.action.VIEW" />
				<category android:name="android.intent.category.DEFAULT" />
				<category android:name="android.intent.category.BROWSABLE" />
				<data android:scheme="msal4a1aa1d5-c567-49d0-ad0b-cd957a47f842" android:host="auth" />
			</intent-filter>
			<intent-filter>
				<action android:name="android.intent.action.VIEW" />
				<category android:name="android.intent.category.DEFAULT" />
				<category android:name="android.intent.category.BROWSABLE" />
				<data android:scheme="msauth" android:host="com.companyname.XamarinDev" android:path="/Fc4l/5I4mMvLnF+l+XopDuQ2gEM=" />
			</intent-filter>
		</activity>
	</application>

```

If you use the above manifest, you can always pass the "msauth://{package name}/{signature hash url encoded}" to the MSAL api as the redirect uri and MSAL will always do the right thing.

Also, as @bgavrilMS mentioned, you need to use the guidance here for creating the broker redirect URI

https://docs.microsoft.com/en-gb/azure/active-directory/develop/msal-net-use-brokers-with-xamarin-apps#step-4-add-a-redirect-uri-to-your-app-registration

"If for discover you have the wrong redirect URI and you get the message. Authentication Error [Android broker] The broker redirect URI is incorrect, it should be msauth://com.xxxxxx.xxxxxxx/xxxxxxxxxxxxxx"

This implies the signature on the signed APK sent to the play store does not match what was used in the development environment. The easiest solution would be to register the URI you get back from the error in the app registration and use that in your code.


You can also extract the package signature like we do in this method here so you dont need to hardcode the redirect URI in your app. Just be sure to have the redirect URI with the signature for both the development env and the playstore registered in the application registration in the portal
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/00cbad0379d232970b95f7871b70892085796546/src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs#L197

Please let me know if you get this to work.

**Comment by ryanwilson-gpdgroup** [comment]
@trwalke Ok. So we were able to successfully authenticate, but now we are facing a problem where any other person who installs the application gets an error saying "Broker Error: User or Admin consent required....", we saw in Azure portal that my name was listed under user consent and I was able to successfully login, but no one else's name would show up. We removed my name from the azure portal and now I get that error as well. We tried granting admin consent for all users in Azure AD but that seems to have no affect. So now we have gone from one issue to another.
